### PR TITLE
Document sw lifetime workaround

### PR DESCRIPTION
**Pull request by sebastianbenz**

### CODE REVIEW

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 1]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+  } finally {
+    clearInterval(keepAlive);
+  }
+}
+
+waitUntil(someExpensiveCalculation());
+```
+
+{% Aside 'important' %}
+An official API simmilar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on Github](https://github.com/w3c/webextensions/issues/416).
```
Comment (oliverdunk) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:278]:
```suggestion
An official API similar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 2]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+{% endAside %}
+
+### Keeping a service worker alive continously
```
Comment (oliverdunk) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:281]:
```suggestion
### Keeping a service worker alive continuously
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 3]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+In rare cases, it is necessary to extend the lifetime indefinitely. Enterprise and EDU are the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future.
```
Comment (oliverdunk) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:283]:
```suggestion
In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and EDU as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 4]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+{% Aside 'important' %}
+An official API simmilar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on Github](https://github.com/w3c/webextensions/issues/416).
+{% endAside %}
+
+### Keeping a service worker alive continously
+
+In rare cases, it is necessary to extend the lifetime indefinitely. Enterprise and EDU are the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
+
+Use the following code snippet to keep your service worker alive as long as your :
```
Comment (oliverdunk) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:285]:
```suggestion
Use the following code snippet to keep your service worker alive:
```

**Review by oliverdunk — APPROVED** [review]
LGTM with nits 👍

File: site/en/docs/extensions/migrating/to-service-workers/index.md (hunk 3 above)
Comment (oliverdunk) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:283]: Alternative suggestion:

```suggestion
It is sometimes desirable to extend the lifetime indefinitely. For extensions intended to be used
for enterprise or EDU use cases on managed devices, keeping a service worker alive can be achieved
by periodically calling a trivial extension API. Beyond this, we recommend the previous approach and
may take action in the future on extensions keeping their service worker alive after any work has
finished.
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 5]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+## Keeping a service worker alive
+
+Service workers are by definition event-driven and will terminate on inactivity. This way Chrome can optimize performance and memory consumption of your extension. Learn more in our [service worker lifecycle documentation](/docs/extensions/mv3/service_workers/service-worker-lifecycle/#idle-shutdown). Exceptional cases might require additional measures to ensure that a service worker stays alive for a longer time.
+
+### Keeping a service worker alive until a long-running operation is finished
+
+When executing a long running operation that does not call any extension APIs in a service worker, the service worker might get shut down while the operation is running. Examples for when this might be needed are:
+
+* A [`fetch` request](https://developer.mozilla.org/docs/Web/API/Fetch_API) potentially taking longer than longer than 5 minutes (e.g. a large download on a potentially poor connection).
```
[the remark this answers is not in the document]
Comment (jpmedley) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:256]: I know the next sentence leaves the digits intact. The standard rules around English numbers are just weird that way.

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 6]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+When executing a long running operation that does not call any extension APIs in a service worker, the service worker might get shut down while the operation is running. Examples for when this might be needed are:
+
+* A [`fetch` request](https://developer.mozilla.org/docs/Web/API/Fetch_API) potentially taking longer than longer than 5 minutes (e.g. a large download on a potentially poor connection). 
+* A complex asynchronous calculation taking more than 30 seconds.
+
+To extend the service worker lifetime in these cases, you can periodically call a trivial extension API to reset the timeout counter. 
+Please note, that this is only reserved for exceptional cases and in most situations there is usually a better, platform idiomatic, way to achieve the same result.
+
+The following code snippet shows a `waitUntil` helper function that will keep your service worker alive until a given promise resolves:
```
Comment (jpmedley) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:262]:
```suggestion
The following example shows a `waitUntil()` helper function that keeps your service worker alive until a given promise resolves:
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 7]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+  } finally {
+    clearInterval(keepAlive);
+  }
+}
+
+waitUntil(someExpensiveCalculation());
+```
+
+{% Aside 'important' %}
+An official API similar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).
```
Comment (jpmedley) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:278]:
```suggestion
An official API similar to `waitUntil()` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 8]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+{% endAside %}
+
+### Keeping a service worker alive continuously
```
Comment (jpmedley) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:281]:
```suggestion
### Keep a service worker alive continuously
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 9]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and EDU as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future.
```
Comment (jpmedley) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:283]:
```suggestion
In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and education as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or education use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
```

File: site/en/docs/extensions/migrating/to-service-workers/index.md [hunk 10]
```diff
@@ -244,3 +244,82 @@ chrome.alarms.onAlarm.addListener(() => {
...
+
+In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and EDU as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
+
+Use the following code snippet to keep your service worker alive:
+
+```js
+/**
+ * Tracks when a service worker was last alive and extends the service worker
+ * lifetime by writing the current time to extension storage every 20 seconds.
+ * You should still prepare for unexpected termination - for example, if the
```
Comment (jpmedley) [inline site/en/docs/extensions/migrating/to-service-workers/index.md:291]: This seems like something you would cover in an intro paragraph.

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: site/en/docs/extensions/migrating/to-service-workers/index.md (hunk 10 above)
Comment (sebastianbenz) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:291]: Yeah, but as it's intended to be copy pasted I thought it's good to leave this in the code.

Comment (oliverdunk) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:291]: To add to this, we expect people to copy paste this without looking at it so adding some detailed comments is a way to hopefully make them read it before they commit it.

**Comment by sebastianbenz** [comment]
@jpmedley addressed the feedback PTAL

**Comment by sebastianbenz** [comment]
argh messed up commits, will update

**Comment by sebastianbenz** [comment]
OK, commit history is fixed, link added to checklist. @jpmedley PTAL

File: site/en/docs/extensions/migrating/to-service-workers/index.md (hunk 10 above)
Comment (jpmedley) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:291]: If developers aren't going to read it, I don't think where you put it makes much of a differences. Frankly, the comments as they are now, don't give me much of a sense of what this is doing. Writing needs to tell me what's going on. A proper overview, which would contain much of the comment material would do that. 

In other words, the explanations are piecemeal and don't explain the big picture. That big picture would have to include enough of the piecemeal information to make the many of the code comments redundant.

Comment (sebastianbenz) [reply site/en/docs/extensions/migrating/to-service-workers/index.md:291]: This looks fine as code comments. These comments are not for developers to be read, but to be copy pasted into the code base.

**Review by jpmedley — APPROVED** [review]

**Pull request merged**